cocotb/drivers/amba.py

```diff
@@ -60,153 +89,401 @@ def __init__(self, entity, name, clock, **kwargs):
+        if size is None:
+            size = len(self.bus.WDATA) // 8
+        else:
+            AXI4Master._check_size(size, len(self.bus.WDATA) // 8)
 
-        await RisingEdge(self.clock)
+        AXI4Master._check_length(len(value), burst)
+        AXI4Master._check_4kB_boundary_crossing(address, burst, size,
+                                                len(value))
 
-        if int(result):
-            raise AXIProtocolError("Write to address 0x%08x failed with BRESP: %d"
-                                   % (address, int(result)))
+        write_address = self._send_write_address(address, len(value), burst,
+                                                 size, address_latency, sync)
 
-        return result
+        write_data = self._send_write_data(value, data_latency, byte_enable,
+                                           sync)
+
+        await Combine(cocotb.fork(write_address), cocotb.fork(write_data))
+
+        async with self.write_response_busy:
+            # Wait for the response
+            while True:
+                await ReadOnly()
+                if self.bus.BVALID.value and self.bus.BREADY.value:
+                    result = AXIxRESP(self.bus.BRESP.value.integer)
+                    break
+                await RisingEdge(self.clock)
+
+            await RisingEdge(self.clock)
+
+            if result is not AXIxRESP.OKAY:
+                err_msg = "Write to address {0:#x}"
+                if len(value) != 1:
+                    err_msg += " ({1} beats, {2} burst)"
+                err_msg += " failed with BRESP: {3} ({4})"
+
+                raise AXIProtocolError(
+                    err_msg.format(address, len(value), burst.name,
+                                   result.value, result.name), result)
 
     @cocotb.coroutine
-    async def read(self, address: int, sync: bool = True) -> BinaryValue:
+    async def read(
+        self, address: int, length: int = 1, *,
+        size: Optional[int] = None, burst: AXIBurst = AXIBurst.INCR,
+        return_rresp: bool = False, sync: bool = True
+    ) -> Union[List[BinaryValue], List[Tuple[BinaryValue, AXIxRESP]]]:
         """Read from an address.
 
         Args:
             address: The address to read from.
-            sync: Wait for rising edge on clock initially.
-                Defaults to True.
+            length: Number of words to transfer. Defaults to 1.
+            size: The size (in bytes) of each beat. Defaults to None (width of
+                the data bus).
+            burst: The burst type, either ``FIXED``, ``INCR`` or ``WRAP``.
+                Defaults to ``INCR``.
+            return_rresp: Return the list of RRESP values, instead of raising a
+                AXIProtocolError in case of not OKAY. Defaults to False.
+            sync: Wait for rising edge on clock initially. Defaults to True.
 
         Returns:
-            The read data value.
+            The read data values or, if **return_rresp** is True, a list of
```

Comment:
```suggestion
            The read data values or, if *return_rresp* is True, a list of
```
Single asterisks (i.e. italics) is convention.